File: be1-go/message/messagedata/federation_result.go
package messagedata

import "popstellar/message/query/method/message"

// FederationResult defines a message data
type FederationResult struct {
	Object string `json:"object"`
	Action string `json:"action"`
	Status string `json:"status"`

	Reason string `json:"reason,omitempty"`

	PublicKey    string          `json:"public_key,omitempty"`
	ChallengeMsg message.Message `json:"challenge"`
}

// GetObject implements MessageData
func (FederationResult) GetObject() string {
	return FederationObject
}

// GetAction implements MessageData
func (FederationResult) GetAction() string {
	return FederationActionResult
}

// NewEmpty implements MessageData
func (FederationResult) NewEmpty() MessageData {
	return &FederationResult{}
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/federation/FederationResult.scala
package ch.epfl.pop.model.network.method.message.data.federation

import ch.epfl.pop.json.MessageDataProtocol.FederationResultFormat
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.Message
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.PublicKey
import spray.json.*

final case class FederationResult(
    status: String,
    reason: Option[String],
    publicKey: Option[PublicKey],
    challenge: Message
) extends MessageData {
  def this(status: String, publicKey: PublicKey, challenge: Message) = this(status, None, Some(publicKey), challenge)
  def this(status: String, reason: String, challenge: Message) = this(status, Some(reason), None, challenge)
  override val _object: ObjectType = ObjectType.federation
  override val action: ActionType = ActionType.result
}

object FederationResult extends Parsable {
  def apply(status: String, reason: Option[String], publicKey: Option[PublicKey], challenge: Message): FederationResult = {
    new FederationResult(status, reason, publicKey, challenge)
  }
  def apply(status: String, publicKey: PublicKey, challenge: Message): FederationResult = new FederationResult(status, None, Some(publicKey), challenge)
  def apply(status: String, reason: String, challenge: Message): FederationResult = new FederationResult(status, Some(reason), None, challenge)

  override def buildFromJson(payload: String): FederationResult = payload.parseJson.asJsObject.convertTo[FederationResult]
}
